packages/lib/storage/src/new/index.ts

```diff
@@ -0,0 +1,143 @@
+ *
+ * @param {Db} db - The database object to interact with.
+ * @returns {(contactEnsName: string, envelop: Envelop) => Promise<void>} A function that takes a contactEnsName string
+ *  and an Envelop object, and performs the operations.
+ */
+function addMessageSideEffectContainment(
+    db: Db,
+): (contactEnsName: string, envelop: Envelop) => Promise<void> {
+    return async (contactEnsName: string, envelop: Envelop) => {
+        const messageChunkContainer = await addMessage(
+            contactEnsName,
+            envelop,
+            db,
+        );
+        if (messageChunkContainer) {
+            await db.keyValueStoreLocal.write(
+                messageChunkContainer.messageChunk.key,
+                messageChunkContainer.messageChunk,
+            );
+            await db.keyValueStoreLocal.write(
+                messageChunkContainer.conversationManifest.key,
+                messageChunkContainer.conversationManifest,
+            );
+            if (
+                messageChunkContainer.messageChunk.envelops.length ===
+                MAX_MESSAGES_PER_CHUNK
+            ) {
+                await db.keyValueStoreLocal.write(
+                    messageChunkContainer.messageChunk.key,
+                    messageChunkContainer.messageChunk,
+                );
+            }
+        }
+    };
+}
+
+/**
+ * This function creates a storage API for managing conversations and messages.
+ * It sets up a local key-value store and a database object,
+ * and returns an API with methods for getting and adding conversations and messages.
+ *
+ * @param {string} accountEnsName - The ENS name of the account.
+ * @param {(data: string) => Promise<string>} sign - A function for signing data.
+ * @param {object} options - Optional configuration options.
+ * @param {ReadStrategy} options.readStrategy - The strategy to use for reading data.
+ * @param {KeyValueStore} options.keyValueStoreRemote - A remote key-value store.
+ * @param {Encryption} options.encryption - An encryption object.
+ *
+ * @returns {StorageAPI} An API with methods for getting and adding conversations and messages.
+ */
+export function createStorage(
+    accountEnsName: string,
+    sign: (data: string) => Promise<string>,
+    options?: Partial<{
+        readStrategy: ReadStrategy;
+        keyValueStoreRemote?: KeyValueStore;
+        encryption?: Encryption;
+    }>,
+): StorageAPI {
+    const keyValueStoreLocal = createKeyValueStore();
+    const db: Db = {
+        readStrategy: ReadStrategy.LocalFirst,
+        accountEnsName,
+        keyValueStoreLocal,
+        sign,
+        // If we read from remote because a chunk is not available, we need to update local storage
+        updateLocalStorageOnRemoteRead: options?.keyValueStoreRemote
+            ? async <T extends Chunk>(key: string, value: T) => {
+                  await keyValueStoreLocal.write(key, value);
+              }
+            : undefined,
```

Comment:
would it make more sense to return a function here to prevent null pointer exceptions when trying to call a function on undefined?